pkg/kubelet/kubelet.go

```diff
@@ -402,66 +406,103 @@ func NewMainKubelet(
 	klet.podCache = kubecontainer.NewCache()
 	klet.podManager = kubepod.NewBasicPodManager(kubepod.NewBasicMirrorClient(klet.kubeClient))
 
-	// Initialize the runtime.
-	switch containerRuntime {
-	case "docker":
-		// Only supported one for now, continue.
-		klet.containerRuntime = dockertools.NewDockerManager(
-			dockerClient,
+	if kubeRuntimeEndpoint != "" {
+		// kubeImageServiceEndpoint is same as kubeRuntimeEndpoint if not specified explicitly
+		if kubeImageServiceEndpoint == "" {
+			kubeImageServiceEndpoint = kubeRuntimeEndpoint
+		}
+
+		// Initialize remote runtime service and image service
+		remoteRuntime, err := remote.NewRemoteRuntime(kubeRuntimeEndpoint, runtimeRequestTimeout)
+		if err != nil {
+			glog.Errorf("Initialize kube runtime failed: %v", err)
+			return nil, err
+		}
+		remoteImage, err := remote.NewRemoteImage(kubeImageServiceEndpoint, runtimeRequestTimeout)
+		if err != nil {
+			glog.Errorf("Initialize kube image service failed: %v", err)
+			return nil, err
+		}
+
+		klet.containerRuntime, err = kuberuntime.NewKubeGenericRuntimeManager(
 			kubecontainer.FilterEventRecorder(recorder),
 			klet.livenessManager,
 			containerRefManager,
-			klet.podManager,
-			machineInfo,
-			podInfraContainerImage,
-			pullQPS,
-			pullBurst,
-			containerLogsDir,
 			osInterface,
 			klet.networkPlugin,
 			klet,
 			klet.httpClient,
-			dockerExecHandler,
-			oomAdjuster,
-			procFs,
-			klet.cpuCFSQuota,
-			imageBackOff,
-			serializeImagePulls,
-			enableCustomMetrics,
-			klet.hairpinMode == componentconfig.HairpinVeth,
-			seccompProfileRoot,
-			containerRuntimeOptions...,
-		)
-	case "rkt":
-		// TODO: Include hairpin mode settings in rkt?
-		conf := &rkt.Config{
-			Path:            rktPath,
-			Stage1Image:     rktStage1Image,
-			InsecureOptions: "image,ondisk",
-		}
-		rktRuntime, err := rkt.New(
-			rktAPIEndpoint,
-			conf,
-			klet,
-			recorder,
-			containerRefManager,
-			klet.podManager,
-			klet.livenessManager,
-			klet.httpClient,
-			klet.networkPlugin,
-			klet.hairpinMode == componentconfig.HairpinVeth,
-			utilexec.New(),
-			kubecontainer.RealOS{},
 			imageBackOff,
 			serializeImagePulls,
-			runtimeRequestTimeout,
+			remoteRuntime,
+			remoteImage,
 		)
 		if err != nil {
+			glog.Errorf("Initialize kube runtime manager failed: %v", err)
 			return nil, err
 		}
-		klet.containerRuntime = rktRuntime
-	default:
-		return nil, fmt.Errorf("unsupported container runtime %q specified", containerRuntime)
+	} else {
```

Comment:
We can save a level of indentation and possibly improve readability by setting `containerRuntime = "remote"` above, removing the else, and moving `NewKubeGenericRuntimeManager` for the remote bit into its own case.

Alternately, if we're not going to be switching entirely based on `containerRuntime` for the construction, we could maybe move to just having an `if..else` ladder that checks if the endpoint is set or the runtime is one of the known ones, and likewise save a level of indentation. I think this isn't as readable as the first suggestion though.